driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java

```diff
@@ -279,17 +277,34 @@ public FieldNameValidator getFieldNameValidator() {
 
     private BulkWriteResult getBulkWriteResult(final BsonDocument result) {
         int count = result.getNumber("n").intValue();
+        List<BulkWriteInsert> insertedItems = getInsertedItems(result);
         List<BulkWriteUpsert> upsertedItems = getUpsertedItems(result);
-        return BulkWriteResult.acknowledged(batchType, count - upsertedItems.size(), getModifiedCount(result), upsertedItems);
+        return BulkWriteResult.acknowledged(batchType, count - upsertedItems.size(), getModifiedCount(result), upsertedItems,
+                insertedItems);
     }
 
-    @SuppressWarnings("unchecked")
+    private List<BulkWriteInsert> getInsertedItems(final BsonDocument result) {
+        if (payload.getPayloadType() == SplittablePayload.Type.INSERT) {
+
+            Stream<WriteRequestWithIndex> writeRequests = payload.getWriteRequestWithIndexes().stream();
+            List<Integer> writeErrors = getWriteErrors(result).stream().map(BulkWriteError::getIndex).collect(Collectors.toList());
+            if (!writeErrors.isEmpty()) {
+                writeRequests = writeRequests.filter(wr -> !writeErrors.contains(wr.getIndex()));
+            }
+            return writeRequests
+                    .map(wr -> new BulkWriteInsert(wr.getIndex(), ((InsertRequest) wr.getWriteRequest()).getId()))
```

Comment:
Here we create the `BulkWriteInsert` list using the modified `InsertRequest` which now contains the `_id`.